exonum/src/runtime/rust/api.rs

```diff
@@ -160,10 +161,87 @@ impl ServiceApiScope {
+        &mut self,
+        name: &'static str,
+        deprecates_on: Option<chrono::Date<chrono::Utc>>,
+        handler: F,
+    ) -> &mut Self
+    where
+        Q: DeserializeOwned + 'static,
+        I: Serialize + 'static,
+        F: Fn(&ServiceApiState<'_>, Q) -> R + 'static + Clone + Send + Sync,
+        R: IntoFuture<Item = I, Error = crate::api::Error> + 'static,
+    {
+        let blockchain = self.blockchain.clone();
+        let descriptor = self.descriptor.clone();
+        let handler = move |query: Q| -> crate::api::FutureResult<I> {
+            let state = ServiceApiState::from_api_context(
+                &blockchain,
+                InstanceDescriptor {
+                    id: descriptor.0,
+                    name: descriptor.1.as_ref(),
+                },
+            );
+            let result = handler(&state, query);
+            Box::new(result.into_future())
+        };
+        // Mark endpoint as deprecated.
+        let mut handler = Deprecated::from(handler);
+        if let Some(deprecates_on) = deprecates_on {
+            handler = handler.with_date(deprecates_on)
+        };
+        self.inner.endpoint(name, handler);
+        self
+    }
+
+    /// Same as `endpoint_mut`, but the response will contain a warning about endpoint being deprecated.
+    /// Optional endpoint expiration date can be included in the warning.
+    pub fn deprecated_endpoint_mut<Q, I, F, R>(
+        &mut self,
+        name: &'static str,
+        deprecates_on: Option<chrono::Date<chrono::Utc>>,
+        handler: F,
+    ) -> &mut Self
+    where
+        Q: DeserializeOwned + 'static,
+        I: Serialize + 'static,
+        F: Fn(&ServiceApiState<'_>, Q) -> R + 'static + Clone + Send + Sync,
+        R: IntoFuture<Item = I, Error = crate::api::Error> + 'static,
+    {
+        let blockchain = self.blockchain.clone();
+        let descriptor = self.descriptor.clone();
+        let handler = move |query: Q| -> crate::api::FutureResult<I> {
+            let state = ServiceApiState::from_api_context(
+                &blockchain,
+                InstanceDescriptor {
+                    id: descriptor.0,
+                    name: descriptor.1.as_ref(),
+                },
+            );
+            let result = handler(&state, query);
+            Box::new(result.into_future())
+        };
+        // Mark endpoint as deprecated.
+        let mut handler = Deprecated::from(handler);
+        if let Some(deprecates_on) = deprecates_on {
+            handler = handler.with_date(deprecates_on)
+        };
+        self.inner.endpoint_mut(name, handler);
+        self
+    }
+
     /// Return a mutable reference to the underlying web backend.
     pub fn web_backend(&mut self) -> &mut crate::api::backends::actix::ApiBuilder {
         self.inner.web_backend()
     }
+
+    /// Returns the base for service URLs, e.g. "/api/services/xnm-token/".
+    pub fn url_base(&self) -> String {
```

Comment:
IMO, a better alternative to exposing this method is to prepend the URL base automatically to the URL returned by the user. Or do you think there is a valid case where the service would redirect to API of another service?

Reply:
I thought that maybe some service will implement a frontend for the network, and instead of direct communication with some node it will be recommended to interact with a server (which will proxy requests to the network). Or is it unrealistic?

Reply:
Jeez, what happens if a node is behind a proxy in general? The location may just be incorrect in this case since the base URL may be incorrect because the proxy layer may add a prefix to node API URLs. Using a relative `Location` URL in this case may make more sense, but requires more work from the service developer.

Reply:
I was just answering the question why I did it that way; no need for anger.

I've tried to implement suggested approach, but faced a problem: `MovedPermanently` endpoins are now the same as usual endpoints, and we can't distinguish the returned result until the handler is invoked (within `actix.rs`). As a result, I wasn't able to implement this easily. The only solution I found is to include the service name into the error, and do prepending when converting `api::Error` into `HttpResponse`, but it looks like a kludge :pensive:

Reply:
No anger implied, sorry. The expression I was going for was "oh my god, it's really difficult to account for the reverse proxy use case, isn't it?"

By using relative URL I meant literally using something like

```rust
api::Error::MovedPermanently("../v2/improved-endpoint")
```

since `Location` HTTP header allows relative URLs. This, of course, would shift the problem to service developers; they would need to compute the number of `../`s correctly.

Reply:
Ah, I get it now.

Hm, and what if we will provide a helper function like `redirect_link(from, to)`, which will produce a relative link from old URL to the new one? It seems that it's pretty simple for implementation, but it will free service developers from counting `..`s.

Reply:
Added a `relative_to` method instead of `url_base`.